sklearn/datasets/openml.py

```diff
@@ -0,0 +1,468 @@
+import json
+import os
+from os.path import join, exists
+from warnings import warn
+
+try:
+    # Python 3+
+    from urllib.request import urlopen
+except ImportError:
+    # Python 2
+    from urllib2 import urlopen
+
+
+import numpy as np
+import scipy.sparse
+
+from sklearn.externals import arff
+from .base import get_data_home
+from ..externals.six import string_types, PY2
+from ..externals.six.moves.urllib.error import HTTPError
+from ..utils import Bunch, Memory
+
+__all__ = ['fetch_openml']
+
+
+_SEARCH_NAME = "https://openml.org/api/v1/json/data/list/data_name/{}/limit/1"
+_DATA_INFO = "https://openml.org/api/v1/json/data/{}"
+_DATA_FEATURES = "https://openml.org/api/v1/json/data/features/{}"
+_DATA_FILE = "https://openml.org/data/v1/download/{}"
+
+
+def _get_json_content_from_openml_api(url, error_message, raise_if_error):
+    """
+    Loads json data from the openml api
+
+    Parameters
+    ----------
+    url : str
+        The URL to load from. Should be an official OpenML endpoint
+
+    error_message : str or None
+        The error message to raise if an acceptable OpenML error is thrown
+        (acceptable error is, e.g., data id not found. Other errors, like 404's
+        will throw the native error message)
+
+    raise_if_error : bool
+        Whether to raise an error if OpenML returns an acceptable error (e.g.,
+        date not found). If this argument is set to False, a None is returned
+        in case of acceptable errors. Note that all other errors (e.g., 404)
+        will still be raised as normal.
+
+    Returns
+    -------
+    json_data : json or None
+        the json result from the OpenML server if the call was successful;
+        None otherwise iff raise_if_error was set to False and the error was
+        ``acceptable``
+    """
+    data_found = True
+    try:
+        response = urlopen(url)
+    except HTTPError as error:
+        # 412 is an OpenML specific error code, indicating a generic error
+        # (e.g., data not found)
+        if error.code == 412:
+            data_found = False
+        else:
+            raise error
+    if not data_found:
+        # not in except for nicer traceback
+        if raise_if_error:
+            raise ValueError(error_message)
+        else:
+            return None
+    json_data = json.loads(response.read().decode("utf-8"))
```

Comment:
shouldn't `json.load` work here without the read/decode?

Reply:
json.loads directly calls to json.load so this shouldn't make any difference? I am fine with changing it though

Reply:
You should use the best API available, rather than assuming things on the basis of the current implementation.

Reply:
Doesn't work for Python 3.X, as the following code `json_data = json.load(response)` raises the following error:
`TypeError: the JSON object must be str, not 'bytes'`

Am I missing something obvious? (There is no way to tell urlopen and gzip open which encoding was used. )

Reply:
json.load has an encoding parameter

Reply:
[The Python 3 json documentation](https://docs.python.org/3/library/json.html) seems to disagree. As a side note, loads does have an encoding parameter, but it is deprecated.